### ISSUE TITLE
Use of liveQuery with infinite scrolling of post

### ISSUE THREAD

**Issue by yatharth77** [dfahlander/Dexie.js#1308]
Hi, 
I am implementing infinite scroll feature in my website(mailing website). My data is stored in indexedDb, in decreasing order of timestamp. I was able to implement infinite scroll, when in case DB does not contain any more mail, I fetch from the api and write in DB which indirectly re-runs the query and appends fetched data at end of current data)

But considering the case, any new mail appears at the top of my DB table. In this case I would want to show it at top in frontend also without re-rendering all the mail.

Can liveQuery help me in any way in this case?

**Comment by dfahlander** [comment]
If I understand you correctly, you want to fetch data and update IndexedDB on scrolling below some point.

* If no new email has arrived, the fetched data would be older emails that are being inserted to IDB.
* If a new email has arrived, all emails would pop a step down and the new email should show highest up.

I think this use case would fit very well with liveQuery. Don't know what UI library you are using, but let me assume you are using React. If so:

* Use `useLiveQuery()` and provide a query that will sort the contents of the db by date and render each email.
  ```js
  const emails = useLiveQuery(() => db.emails.orderBy('date').limit(1000))
  ```
* When you detect scroll position comes at a certain treshold, launch your fetch action and let it just write to the db.
* Your component will update correctly the fetched data has landed in IndexedDB and react would do the diff to prohibit unnescessary re-rendering of existing emails.

This would scale up to a certain point. Use React.memo() to optimize each react component representing a rendered email - if all props are equal, react will skip internal re-rendering as well as the physical dom re-rendering.

If you would want to scale this for tens of thousands emails, I think you would need to do some heavy lifting yourself to workaround limitations in all levels: browser, React and IndexedDB. You might not want to physically render all emails, but render fake content outside the visible part of in the scrolled area. Check if there are other helper libraries for this. If so, you could still be using liveQuery but just add some `fromDate` parameter to your query, which would need to be computed some way.

```js
const emails = useLiveQuery(() =>
  db.emails.where('date').below(fromDate).limit(1000) // will also be sorted by date
)
```

**Comment by yatharth77** [comment]
Yes, you understood quite well and yes I am working on react. Still one doubt remains with me.

Firstly, my view will fetch 30-40 mails at a time (not 1000 mails). I was thinking to use limit and offset.
In beginning my offset will be 0 and limit would be 40, first 40 mails (0 to 39) will be fetched and rendered
Then offset will be change to 40 and limit will remain 40, now mails from 40 to 79 will be fetched (which may not be present in idb, in that case, I will fetch from API) and append at end of 0-39 mail list (when my curse reaches to certain point in the screen).

Now, doubt is: while fetching newer mail and what if a new mails come at top of my table. I want to push it at top of 0-39 mails rendered already. So now I am able to run live query for mails appending below but how do I handle mails which needs to be added in the beginning.

The problem with fetching all mails from 0 every-time would be unnecessary overhead, so I want to fetch the next mails only. But also want to manage insertion of newer mail.

**Comment by yatharth77** [comment]
Thank you for replying in detail. Also apart form it, I wanted to ask where I can find to read detailed working of liveQuery(). Like the maximum amount of explanation. I was curios about knowing the internal working of it.

**Comment by dfahlander** [comment]
# useLiveQuery() with offset and limit

Using offset/limit is also fine. As shown in some of my samples, you can use a react prop or state in your query, such as the offset and the limit. Just remember to put these variables in the second argument to useLiveQuery() so it will update when the offset or limit is changed.

What will happen is this: Say you are monitoring all emails from offset 100 with limit 40

1. Query is being launched. Your component gets undefined as result and you handle that.
2. Query response for items 100...139 arrives. Your component is rendered.
3. A new email arrives that would be sorted first of all emails. Query will relaunch and return current items 100..139 which will be different than before. The one that was the first email previously will now be the second. 

Maybe this is what you also expect. If not, you could also keep the date of the previous page in a state and when loading the next page, do a date based query as I examplified in my previous comment. If going for this paging approach, the view would not change if a new email arrives, only when user scrolls up to the top.

# The Internals of liveQuery

Regarding the internal details of liveQuery, I haven't written them down in detail yet. But the principles are:

* Every query within a liveQuery context tracks the queried ranges on every queried index and listens to the txcommitted event.
* On every successfully committed write-transaction, an event txcommitted is fired, containing all ranges on indexes and primary keys that are affected.
* The txcommitted event is also propagated across windows and workers
* Queriers compare their collected range of queried indexes against the mutated once. If any range overlap, the querier will zero out all subscriptions and relaunch to get a fresh updated result and also start subscribing again to the ranges that will be accurate this run.

To accomplish range overlap checks in O(n) and avoid loops in loops, there's a class RangeSet that is a binary tree of ranges.
On top of this there are som optimisations such as:
* If more than 50 rows are added to indexedDB, skip tracking ranges and regard all indexes and keys as changed. This optimisation is for not slowing down bulkAdds of tens of thousands items. False positives are cheaper in this case. The limit of 50 is not scientifically the most optimal but probably a decent treshold. 
* We cannot always know the existing values of the indexes. We could do a transactional bulkGet() under the hood to figure them out, but I chosed to not slow down puts but rather do some tricks to overcome it instead. 

If curious about how it works, the code for this lies in [src/live-query](https://github.com/dfahlander/Dexie.js/tree/master/src/live-query)

**Comment by scazzy** [comment]
Extending `RangeSet` api in https://github.com/dfahlander/Dexie.js/blob/master/src/helpers/rangeset.ts

@dfahlander my use case is - I've a virtual list implementation for `0...Infinite` data. Planning to observe on dexie query for the infinite scrolling list.

While I am able to observe on a set query with a specific rangeset, what I'm looking for is to scale this a li'l bit while still keeping the query and the resultset light.
Like a traditional offset-limit mechanism, I want to have a observable on a query, and on scroll I want to change the rangeset in the observable while keeping the query intact.

`var observer = livequery(query({cursor: 1, limit: 10}));`
on scroll, the query remains the same, and you want to fetch next set of data with just the offset/cursor being changed. 
`observer.updateRange({cursor:11,limit:10});`

This will help avoid fetching abundant data in memory, and avoiding query matching on every data change, while only re-querying when the offset/cursor changes.

**Comment by yatharth77** [comment]
![Screenshot 2021-06-14 at 11 53 57 AM](https://user-images.githubusercontent.com/34445235/121848315-8254e900-cd07-11eb-9116-adc030317ffe.png)

I tried doing an experiment on live Query. If you see the graph, there are 3 curves, red, yellow and blue.

If you see the graph, the caption of the graph has a fraction A/B. This means: Let's say Q1 is a liveQuery, and now due to change in state, the query becomes Q2. The output of Q2 contains B number of objects out of which only A number of object are new. That is B-A were also the part of output of Q1. 

example:
Output of Q1 -> 100 - 200

after Query changes:
Output of Q2 -> 120 - 220

Here 
B = 100 (total number of objects)
A = 20 (from Q1 to Q2, only 20 new objects are required)

**NOTE: ALL THE OBJECTS FETCHED ARE INCREASING ORDER OF INDEX

So assuming this
Red - 20/100 - out of 100 objects, only 20 objects are new
Blue - 100/100 - out of 100 objects, all the 100 objects are new
Yellow - 0/100 - out of 100 objects, 0 objects are new (i.e. I am fetching same set of objects again and again)

So, what I read from the documentation and what you explained in the previous comment, Red curve must have been below the blue one because LiveQuery should have automatically understood that 80 objects are old so, it should not refetch them (using rangeSet). But they both are nearly equal which means even time for 100 new objects is equal to 20 new + 80 old objects.

Also the yellow curve is made when I am fetching the same objects again and again. So according to rangeSet theory, time for refetching after 1st refetch must be very small. But if you see it is a linear graph which means every time live Query is taking same amount of time to fetch same objects again and again.

**Comment by dfahlander** [comment]
Hi! liveQuery() computes whether a mutation would affect the query or not. It does not compute which parts of the result that needs to be refetched. Whenever a certain query is affected by a mutation, it will be refetched in full. I hope the documentation doesn't say otherwise.

**Comment by yatharth77** [comment]
Okey so rangeQuery optimisation works when Query is same but the output of it changes (due to transaction in DB)
In other case when Query is itself changed, the whole range is refetched!! Right?

**Comment by Salman9000** [comment]
Any update on this?
